doc: OData client JS usage (#189)
SAP/cloud-sdk: docs/js/features/odata/use-generated-client.md

BusinessPartnerAddress.requestBuilder().create(partner);
BusinessPartnerAddress.requestBuilder().getAll();
BusinessPartnerAddress.requestBuilder().getByKey('id');
BusinessPartnerAddress.requestBuilder().update(partner);
BusinessPartnerAddress.requestBuilder().delete();
```

Each of the above statements returns a builder object that allows for specifying certain request parameters, depending on the operation.

<!--Maybe the Query parameters and Request parameters are redundant because already explained previously?-->

The following query parameters and request options are available for these operations:

Query parameters:
- `$select` and `$expand` are available on reading a single or multiple entities
- `$filter`, `$top`, `$skip` and `$orderby` are available only when reading a collection of entities

Request parameters:
- Update and delete operations allow to modify how ETags are handled:
   - By default an ETag is send if one is present on the entity being modified.
   - `ignoreVersionIdentifier()` will instead always send a `*` which acts as a wildcard to match all ETags.
- All operations allow for adding custom headers via `withCustomHeaders(...)`

#### Handling of ETags

> An [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) is a version identifier which is often used to implement an optimistic locking mechanism. The SDK will try to read version identifiers from responses and set them when sending OData requests.

Reviewer: ```suggestion
An [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) is a version identifier which is often used to implement an optimistic locking mechanism. The SAP Cloud SDK will try to read version identifiers from responses and set them when sending OData requests.
```

Consider the following example:

```ts

> //the function containing this code must be async
> const partner = BusinessPartner.requestBuilder()
>     .getByKey('id')
>     .execute({
>         url: '<yourURL>'
>     });
> BusinessPartner.requestBuilder()
>     .update(await partner)
>     .execute({
>         url: '<yourURL>'
>     });

Reviewer: I would remove that comment. Instead you can consider to just wrap it in an async function. 
```suggestion
async function modifyBusinessPartner(id) {
  const destination = { url: 'https://my.s4-system.com' };
  
  const partner = await BusinessPartner.requestBuilder()
      .getByKey(id)
      .execute(destination);
      
  // do some modification
  applyModification(partner);
  
  return BusinessPartner.requestBuilder()
      .update(partner)
      .execute(destination);
}
```

```

> On the read request the SDK will automatically try to extract the version identifier from the response and store it within the `partner` object. When updating it will be taken from there and sent with the `If-match` header.

Reviewer: ```suggestion
When executing `getAll` and `getByKey` requests the SAP Cloud SDK will automatically attempt to extract the version identifier from the response and store it within the returned entity (`partner` in the example above). When executing `update` requests the version identifier will be sent in the `If-match` request header.
```

The [OData v4 standard](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_SystemQueryOptionfilter) allows for a wide range of filter expressions. A detailed list of what is available in the SDK can be obtained from [the documention](https://sap.github.io/cloud-sdk/docs/js/api-reference-js-ts/). The functionality can also be discovered through the fluent API.

The below example leverages OData v4 exclusive features to build a more complex request:

```ts
/*
Fetch all business partners where:
- the last name is at least twice as long as the first name
- AND the combined string of first and last name does not contain 'bob'
*/
BusinessPartner.requestBuilder()
    .getAll()
    .filter(
        and(
            filterFunction('multiply', 'int', filterFunction('length', 'int', BusinessPartner.FIRST_NAME), '2')
                .lessThan(BusinessPartner.LAST_NAME.edmType.length)
            ,substringOf(
                filterFunction('concat', 'string', BusinessPartner.FIRST_NAME, BusinessPartner.LAST_NAME._fieldName),
                'bob')
                    .equals(true)
        )

>     )

Reviewer: This example is like a hammer on my head. I think this should be split into two smaller examples that we can take a closer look at in the accompanying text:

1. We have predefined fns, e. g. length, substring, substringOf (this can actually even be linked in the documentation - you will have to search for it otherwise you won't find it).
```ts
.filter(length(BusinessPartner.FIRST_NAME).lessThan(5))
```

2. You can construct other filter functions:
```
.filter(filterFunction('multiply', 'int', length(BusinessPartner.FIRST_NAME, '2').lessThan(5)))
```

Also the construction of additional filter functions is not explained in the text. We can also leave out the filter functions for now. and just add this as a comment in the file.

Author: @marikaner I can't use the length function, the types seem to be incompatible but I can't figure out the problem
![image](https://user-images.githubusercontent.com/66725443/85555372-2260c080-b626-11ea-9157-fd4312ee0246.png)

    .getAll()
    .select(
        BusinessPartner.FIRST_NAME,
        BusinessPartner.TO_BUSINESS_PARTNER_ADDRESS.select(
            BusinessPartnerAddress.ADDRESS_ID,
            BusinessPartnerAddress.CITY_CODE
        )
    )
    .execute(destination);
```

The above translates to the following query parameters:

```sql
$select=FirstName,to_BusinessPartnerAddress/AddressID,to_BusinessPartnerAddress/CityCode&$expand=to_BusinessPartnerAddress
```

### Filter

When operating on a collection of entities the API offers `filter( ... )` on the builders. It directly corresponds to the `$filter` parameter of the request. Filters are also build via the static property fields on entities.

The following example:

```ts
/*
Get all business partners that either:
  - Have first name 'Alice' but not last name 'Bob'
  - Or have first name 'Mallory'
*/
BusinessPartner.requestBuilder()
    .getAll()
    .filter(
        or(
            and(
                BusinessPartner.FIRST_NAME.equals('Alice'),
                BusinessPartner.LAST_NAME.notEquals('Bob')
            ),
            BusinessPartner.FIRST_NAME.equals('Mallory')
        )
    )
    .execute(destination);
```

> Will translate to this filter parameter:

Reviewer: ```suggestion
The example above will translate to this filter parameter:
```

```sql
$filter=(((FirstName eq 'Alice') and (LastName ne 'Bob')) or (FirstName eq 'Mallory'))
```

Take note of the order of `and` and `or`. As `or` is invoked on the result of `and` it will form the outer expression while `and` is an inner expression in the first branch of `or`.

> It's also possible to add multiple filters in the same filter function without using `and`, as they will be concatenated. This example:

Reviewer: ```suggestion
It is also possible to pass multiple filters to the same filter function without concatenating them with `and` or `or`. They will be concatenated with `and` by default. The two following examples are equal:
```

```ts
.filter(
    and(
        BusinessPartner.FIRST_NAME.equals('Alice'),
        BusinessPartner.LAST_NAME.notEquals('Bob')
    )
)
```

> Can be reduced to this:

Reviewer: The example above can be shortened to:

```ts
.filter(
    BusinessPartner.FIRST_NAME.equals('Alice'),
    BusinessPartner.LAST_NAME.notEquals('Bob')
)
```

> More advanced filter can be found [here](#Available-Filter-Expressions)

Reviewer: ```suggestion
More advanced filter expressions can be found [here](#Available-Filter-Expressions).
```

> #### OrderBy
> 
> Add order-by statements to the request.

Reviewer: ```suggestion
<!--
#### OrderBy
TBD
Add order-by statements to the request.
-->
```

#### Other Set operations

> - `skip` skips a number of entities. Useful for paging.
> - `top` limits the number of returned entities.

Reviewer: ```suggestion
`skip` allows you to skip a number of results in the requested set. It can be useful for paging:

```ts
MyEntity.requestBuilder().
  .getAll()
  .skip(10)
```

The example above skips the first 10 results.

`top` limits the number of returned results. This can also be useful for paging:
```ts
MyEntity.requestBuilder().
  .getAll()
  .top(10)
```

The example above skips the first 10 results, that match your request.
```

---
id: use-typed-odata-client-for-js
title: Use the type-safe client to consume OData Services for JS/TS
hide_title: false
hide_table_of_contents: false
sidebar_label: Use the type-safe client to consume OData Services for JS/TS
description: Use SAP Cloud SDK for JavaScript to build and run OData requests in a typesafe way.
keywords:
- sap
- cloud
- sdk
- odata
- JavaScript 
- TypeScript
- consume
author: Charles Dubois
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

## Making a request with the generated OData client

The typed OData client allows to build type-safe OData requests for a given service.

> Every OData client generated by the SAP Cloud SDK provides request builders and abstractions for entity sets, that we refer to as entities. In order to create a request towards a specific entity set, leverage the fluent API of an entity.

Reviewer: ```suggestion
Every OData client generated by the SAP Cloud SDK provides request builders and abstractions for entity sets, that we refer to as entities. In order to create a request towards a specific entity set leverage the fluent API of an entity:
```

> First, import the generated service:
> 
> ```ts
> import { MyEntity } from './outputDir/my-service';
> ```
> 
> Where `outputDir` is the directory you specified when generating the service and `my-service` was the ` directoryName` specified in the `serviceMapping.json` file which is inside your input directory.

Reviewer: ```suggestion
```

> On an abstract level, creating a request using the fluent api always follows the same simple structure:

Reviewer: ```suggestion
On an abstract level, creating a request using the fluent API always follows the same simple structure:
```

```ts

> return MyEntity.requestBuilder()
>     .<requestType>(...)
>     .<additionalRequestConfiguration>(...)
>     .execute(destination);

Reviewer: ```suggestion
import { MyEntity } from './outputDir/my-service';

return MyEntity.requestBuilder()
    .<requestType>(...)
    .<additionalRequestConfiguration>(...)
    .execute(destination);
```

```

> 

Reviewer: ```suggestion
First, import the entity from the service directory (`my-service`) within the output directory you specified when generating the service (`outputDir`). The service directory corresponds to the ` directoryName` specified in the `serviceMapping.json` in your input directory.
```

:::

The request can further be configured by chaining additional configuration functions (denoted by `additionalRequestConfiguration` in the example above). All requests can be configured by setting [custom request headers](#Setting-custom-request-headers), [custom query parameters](#Setting-custom-query-parameters) and a [custom service path](#Setting-custom-service-path).

Each request type has additional request specific configuration options:

Creating  an entity `asChildOf` another entity for [`create`](#Create) requests, [ETag handling](#Handling-of-ETags) for [`update`](#Update) and [`delete`](#Delete) requests, as well as set operations for [`getAll`](#GetAll) requests and [`select`](#Select)ing properties for [`getAll`](#GetAll) and [`getByKey`](#GetByKey) requests.
<!--
*3 if there was a specific section on create this should rather be explained there and in more detail.
-->

The last step when making a request using the SAP Cloud SDK is the request execution. Once the request is configured, chain the `execute` function and pass a [destination](/cloud-sdk/docs/js/features/connectivity/destination-js-sdk/) to it.

### Setting custom request headers

> To use the SAP API Business Hub sandbox for your requests, you will need to pass the `apikey` using the `withCustomHeaders` method. You can also set the `authorization` with this method.
> 
> :::caution
> Setting `authorization` (and `apikey`) headers will overwrite any automatic authorization header building that the SDK would normally do.
> :::
> 
> Checkout the following example:
> 
> ```ts
> MyEntity.requestBuilder()
>     .getAll()
>     .withCustomHeaders({
>         apikey: 'my-api-key'
>     });
> ```

Reviewer: ```suggestion
The SAP Cloud SDK automatically sets some necessary request headers on every request. You can specify additional custom headers using the `withCustomHeaders` function:

```ts
MyEntity.requestBuilder()
    .getAll()
    .withCustomHeaders({
        apikey: 'my-api-key'
    });
```

The keys and values of the passed object correspond with the header names and values. Custom headers take priority over automatically generated headers. Automatically generated headers are sent with lower case header names, custom header keep the case they were set in.
In the example above we pass an `apikey`, which is necessary if you want to make use of the SAP API Business Hub sandbox.

:::caution
Setting an `authorization` or `apikey` header (independent of lowercase or uppercase spellint) will skip any automatic authorization header building that the SAP Cloud SDK would normally do.
:::

Checkout the following example:

```

### Setting custom query parameters

> This method allows you to set key-value pairs in the request
> Key-value pairs denoting additional custom query parameters to be set in the request
> 
> :::caution
> Setting custom query parameters overwrites any other configuration that might have been created by the SDK.
> :::
> 
> ```ts
> MyEntity.requestBuilder()
>     .getAll()
>     .withCustomQueryParameters({
>         language: 'en'
>     });

Reviewer: ```suggestion
```suggestion
The SAP Cloud SDK automatically sets some necessary request headers on every request. You can specify additional custom headers using the `withCustomHeaders` function:

```ts
MyEntity.requestBuilder()
    .getAll()
    .withCustomHeaders({
        apikey: 'my-api-key'
    });

```

```

> 

Reviewer: ```suggestion

The keys and values of the passed object correspond with the header names and values. Custom headers take priority over automatically generated headers. Automatically generated headers are sent with lower case header names, custom header keep the case they were set in.
In the example above we pass an `apikey`, which is necessary if you want to make use of the SAP API Business Hub sandbox.

:::caution
Setting an `authorization` or `apikey` header (independent of lowercase or uppercase spellint) will skip any automatic authorization header building that the SAP Cloud SDK would normally do.
:::

```
